main.go

```diff
@@ -263,17 +224,115 @@ func computeRoutePrefix(prefix string, externalURL *url.URL) string {
 
 // shutdownServerOnQuit shutdowns the provided server upon closing the provided
 // quitCh or upon receiving a SIGINT or SIGTERM.
-func shutdownServerOnQuit(server *http.Server, quitCh <-chan struct{}, logger log.Logger) error {
+func shutdownServerOnQuit(socketPath string, options []*network.ServerOption, ms *storage.DiskMetricStore, errCh <-chan error, logger log.Logger) error {
 	notifier := make(chan os.Signal, 1)
 	signal.Notify(notifier, os.Interrupt, syscall.SIGTERM)
 
 	select {
 	case <-notifier:
 		level.Info(logger).Log("msg", "received SIGINT/SIGTERM; exiting gracefully...")
 		break
-	case <-quitCh:
-		level.Warn(logger).Log("msg", "received termination request via web service, exiting gracefully...")
+	case err := <-errCh:
+		level.Warn(logger).Log("msg", "received error when launching server, exiting gracefully...", "err", err)
 		break
 	}
-	return server.Shutdown(context.Background())
+
+	defer os.Remove(socketPath)
+
+	var retErr error
+	for _, option := range options {
+		err := option.Server.Shutdown(context.Background())
+		if err != nil {
+			level.Error(logger).Log("msg", "unable to shutdown the server", "err", err)
+			retErr = errors.Join(retErr, err)
+		}
+	}
+
+	err := ms.Shutdown()
+	if err != nil {
+		level.Error(logger).Log("msg", "unable to shutdown the storage service", "err", err)
+		retErr = errors.Join(retErr, err)
+	}
+	return retErr
+}
+
+// startVerificationServer starts a verification server listening the unix socket
+// it is also responsible for authentication via pid.
+func startVerificationServer(option *network.ServerOption) {
+	level.Info(option.Logger).Log("msg", "Start verification server")
+	unixListener, err := peercred.Listen(context.Background(), option.SocketPath)
+	if err != nil {
+		level.Error(option.Logger).Log("msg", "Could not create local unix socket", "err", err)
+		option.ErrCh <- err
+		return
+	}
+
+	// chmod socketPath
+	err = os.Chmod(option.SocketPath, 0o777)
+	if err != nil {
+		level.Error(option.Logger).Log("msg", "Could not chmod local unix socket", "err", err)
+		option.ErrCh <- err
+		return
+	}
+
+	listener := network.WrappedListener{
+		Listener:     unixListener,
+		TrustedPath:  option.TrustedPath,
+		Option:       option,
+		ContainerMap: make(map[string]*network.WrappedInstance),
+	}
+
+	quitCh := make(chan struct{}, 1)
+
+	go func() {
+		err = web.Serve(&listener, option.Server, option.WebConfig, option.Logger)
+		if err != nil {
+			if errors.Is(err, http.ErrServerClosed) {
+				level.Info(option.Logger).Log("msg", "Verification server stopped")
+			} else {
+				level.Error(option.Logger).Log("msg", "Verification server stopped with error", "err", err)
+				option.ErrCh <- err
+			}
+		}
+
+		quitCh <- struct{}{}
+	}()
+
+	for {
+		for id, info := range listener.ContainerMap {
```

Comment:
Could eventually be refactored a bit with a channel to pass `*network.WrappedInstance` without micro sleeps

Reply:
thanks! I have refactored the code and removed the micro sleeps.